### Wrangle Data

The following data was acquired from the Austin Animal Center and was last updated October 11th, 2021. There are two different files. One is for the animal center intakes and the other is for the animal center outcomes. I plan on merging both of these files so that I have information on which animals were adopted and which weren't. I can't just use the outcomes file because it only includes the animals that had a certain outcome.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

### Get the Intake Data

In [2]:
intakes = pd.read_csv('Austin_Animal_Center_Intakes.csv')

In [3]:
intakes.head()

Animal ID     Name                DateTime               MonthYear  \
0   A786884   *Brock  01/03/2019 04:19:00 PM  01/03/2019 04:19:00 PM   
1   A706918    Belle  07/05/2015 12:59:00 PM  07/05/2015 12:59:00 PM   
2   A724273  Runster  04/14/2016 06:43:00 PM  04/14/2016 06:43:00 PM   
3   A665644      NaN  10/21/2013 07:59:00 AM  10/21/2013 07:59:00 AM   
4   A682524      Rio  06/29/2014 10:38:00 AM  06/29/2014 10:38:00 AM   

                        Found Location Intake Type Intake Condition  \
0  2501 Magin Meadow Dr in Austin (TX)       Stray           Normal   
1     9409 Bluegrass Dr in Austin (TX)       Stray           Normal   
2   2818 Palomino Trail in Austin (TX)       Stray           Normal   
3                          Austin (TX)       Stray             Sick   
4        800 Grove Blvd in Austin (TX)       Stray           Normal   

  Animal Type Sex upon Intake Age upon Intake  \
0         Dog   Neutered Male         2 years   
1         Dog   Spayed Female         8 years   
2         Dog     Intact Male       11 months   
3         Cat   Intact Female         4 weeks   
4         Dog   Neutered Male         4 years   

                                   Breed        Color  
0                             Beagle Mix     Tricolor  
1               English Springer Spaniel  White/Liver  
2                            Basenji Mix  Sable/White  
3                 Domestic Shorthair Mix       Calico  
4  Doberman Pinsch/Australian Cattle Dog     Tan/Gray

In [4]:
intakes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132220 entries, 0 to 132219
Data columns (total 12 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Animal ID         132220 non-null  object
 1   Name              91800 non-null   object
 2   DateTime          132220 non-null  object
 3   MonthYear         132220 non-null  object
 4   Found Location    132220 non-null  object
 5   Intake Type       132220 non-null  object
 6   Intake Condition  132220 non-null  object
 7   Animal Type       132220 non-null  object
 8   Sex upon Intake   132219 non-null  object
 9   Age upon Intake   132220 non-null  object
 10  Breed             132220 non-null  object
 11  Color             132220 non-null  object
dtypes: object(12)
memory usage: 12.1+ MB


Key Takeaways:
* There are 132,220 entries.
* MonthYear appears to be the same info as DateTime, so I can drop it. I think I will create a new column called 'time_at_shelter' and investigate for a relationship with adoptability.
* There are a few null values, but most of them can be dropped without worry.
* I will drop 'Name' because it won't have any bearing on adoptability
* I will also drop 'Found Location' because it shouldn't matter.
* I will drop 'DateTime' after converting it to a datetime object and creating a new column called intake_date.
* I will rename the 'Intake Type', 'Intake Condition', 'Animal Type', 'Sex upon Intake', and 'Age upon Intake' columns
* There is a single null for 'Sex upon Intake'. I will probably just guess the sex, but may end up dropping it.
* The age column can be dropped since I can calculate the age_in_days with the next data set.
* The Sex upon Intake column can be dropped. I will use the same information from the outcomes dataset to create the columns I want.

In [5]:
intakes = intakes.drop(columns = ['Name', 'MonthYear', 'Found Location', 'Age upon Intake', 'Sex upon Intake'])

In [6]:
intakes.rename(columns = {
    'Intake Type': 'intake_type',
    'Intake Condition': 'intake_condition',
    'Animal Type': 'animal_type',
}, inplace = True)

In [7]:
#Assume the animal with a missing sex value is a netured male
intakes = intakes.fillna('Neutered Male')

In [8]:
intakes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132220 entries, 0 to 132219
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Animal ID         132220 non-null  object
 1   DateTime          132220 non-null  object
 2   intake_type       132220 non-null  object
 3   intake_condition  132220 non-null  object
 4   animal_type       132220 non-null  object
 5   Breed             132220 non-null  object
 6   Color             132220 non-null  object
dtypes: object(7)
memory usage: 7.1+ MB


In [9]:
intakes.isnull().sum()

Animal ID           0
DateTime            0
intake_type         0
intake_condition    0
animal_type         0
Breed               0
Color               0
dtype: int64

In [10]:
#Convert 'DateTime' to datetime object
intakes.DateTime = pd.to_datetime(intakes.DateTime)

In [11]:
#Create new column 'intake_date' then drop 'DateTime'
intakes['intake_date'] = pd.to_datetime(intakes.DateTime.dt.date)
intakes = intakes.drop('DateTime', axis = 1)

In [12]:
intakes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132220 entries, 0 to 132219
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   Animal ID         132220 non-null  object        
 1   intake_type       132220 non-null  object        
 2   intake_condition  132220 non-null  object        
 3   animal_type       132220 non-null  object        
 4   Breed             132220 non-null  object        
 5   Color             132220 non-null  object        
 6   intake_date       132220 non-null  datetime64[ns]
dtypes: datetime64[ns](1), object(6)
memory usage: 7.1+ MB


In [13]:
#Now order the dataframe by Animal ID and intake_date.
#Then remove duplicates but leave the latest observation
intakes = intakes.sort_values(by = ['Animal ID', 'intake_date'])

In [14]:
intakes.head()

Animal ID      intake_type intake_condition animal_type  \
84165   A006100    Public Assist           Normal         Dog   
3988    A006100    Public Assist           Normal         Dog   
18633   A006100            Stray           Normal         Dog   
65397   A047759  Owner Surrender           Normal         Dog   
88961   A134067    Public Assist          Injured         Dog   

                      Breed         Color intake_date  
84165  Spinone Italiano Mix  Yellow/White  2014-03-07  
3988   Spinone Italiano Mix  Yellow/White  2014-12-19  
18633  Spinone Italiano Mix  Yellow/White  2017-12-07  
65397             Dachshund      Tricolor  2014-04-02  
88961     Shetland Sheepdog   Brown/White  2013-11-16

In [15]:
#Remove duplicates
intakes = intakes.drop_duplicates(subset = 'Animal ID', keep = 'last', ignore_index = True)

In [16]:
intakes.head()

Animal ID      intake_type intake_condition animal_type  \
0   A006100            Stray           Normal         Dog   
1   A047759  Owner Surrender           Normal         Dog   
2   A134067    Public Assist          Injured         Dog   
3   A141142            Stray             Aged         Dog   
4   A163459            Stray           Normal         Dog   

                         Breed         Color intake_date  
0         Spinone Italiano Mix  Yellow/White  2017-12-07  
1                    Dachshund      Tricolor  2014-04-02  
2            Shetland Sheepdog   Brown/White  2013-11-16  
3  Labrador Retriever/Pit Bull   Black/White  2013-11-16  
4      Miniature Schnauzer Mix    Black/Gray  2014-11-14

In [17]:
intakes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 118225 entries, 0 to 118224
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   Animal ID         118225 non-null  object        
 1   intake_type       118225 non-null  object        
 2   intake_condition  118225 non-null  object        
 3   animal_type       118225 non-null  object        
 4   Breed             118225 non-null  object        
 5   Color             118225 non-null  object        
 6   intake_date       118225 non-null  datetime64[ns]
dtypes: datetime64[ns](1), object(6)
memory usage: 6.3+ MB


### Acquire Outcome Data

In [18]:
outcomes = pd.read_csv('Austin_Animal_Center_Outcomes.csv')

In [19]:
outcomes.head()

Animal ID   Name                DateTime               MonthYear  \
0   A794011  Chunk  05/08/2019 06:20:00 PM  05/08/2019 06:20:00 PM   
1   A776359  Gizmo  07/18/2018 04:02:00 PM  07/18/2018 04:02:00 PM   
2   A821648    NaN  08/16/2020 11:38:00 AM  08/16/2020 11:38:00 AM   
3   A720371  Moose  02/13/2016 05:59:00 PM  02/13/2016 05:59:00 PM   
4   A674754    NaN  03/18/2014 11:47:00 AM  03/18/2014 11:47:00 AM   

  Date of Birth Outcome Type Outcome Subtype Animal Type Sex upon Outcome  \
0    05/02/2017    Rto-Adopt             NaN         Cat    Neutered Male   
1    07/12/2017     Adoption             NaN         Dog    Neutered Male   
2    08/16/2019   Euthanasia             NaN       Other          Unknown   
3    10/08/2015     Adoption             NaN         Dog    Neutered Male   
4    03/12/2014     Transfer         Partner         Cat      Intact Male   

  Age upon Outcome                               Breed              Color  
0          2 years              Domestic Shorthair Mix  Brown Tabby/White  
1           1 year             Chihuahua Shorthair Mix        White/Brown  
2           1 year                             Raccoon               Gray  
3         4 months  Anatol Shepherd/Labrador Retriever               Buff  
4           6 days              Domestic Shorthair Mix       Orange Tabby

In [20]:
outcomes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132327 entries, 0 to 132326
Data columns (total 12 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Animal ID         132327 non-null  object
 1   Name              91926 non-null   object
 2   DateTime          132327 non-null  object
 3   MonthYear         132327 non-null  object
 4   Date of Birth     132327 non-null  object
 5   Outcome Type      132302 non-null  object
 6   Outcome Subtype   60804 non-null   object
 7   Animal Type       132327 non-null  object
 8   Sex upon Outcome  132326 non-null  object
 9   Age upon Outcome  132307 non-null  object
 10  Breed             132327 non-null  object
 11  Color             132327 non-null  object
dtypes: object(12)
memory usage: 12.1+ MB


Key Takeaways:
* There are 127 more entries here than in the intakes dataset for some reason. 
* The Name column can be dropped
* The DateTime column will be dropped after creating a new column called 'outcome_date'. I will also need to use it to calculate time at shelter.
* The MonthYear column can be dropped
* The Date of Birth column will be dropped after creating a new column called 'outcome_age'
* The Outcome type column will be renamed
* I will also drop all rows where the outcome type is transfer, since I cannot tell if these animals were ever adopted or not.
* I only need to know if the animal was adopted or not, so outcome subtype can be dropped.
* The Animal Type column can be dropped since it should be the same as the previous dataset
* The Sex upon Outcome column will be used to set up two new columns. The first will be the sex of the animal, and the second will be whether or not the animal was spayed or neutered.
* The Age upon Outcome column will be dropped
* The Breed and Color columns will be dropped since that information is in the first dataset

In [21]:
#convert DateTime column to datetime object and then create new column 'outcome_date'
outcomes.DateTime = pd.to_datetime(outcomes.DateTime)

In [22]:
outcomes['outcome_date'] = pd.to_datetime(outcomes.DateTime.dt.date)

In [23]:
#convert date of birth to datetime object and then create new column 'age_in_days'
outcomes['Date of Birth'] = pd.to_datetime(outcomes['Date of Birth'])

In [24]:
outcomes['age_in_days'] = (outcomes.outcome_date - outcomes['Date of Birth']) // pd.Timedelta('1d')

In [25]:
outcomes = outcomes.drop(columns = [
    'Name',
    'DateTime',
    'MonthYear',
    'Date of Birth',
    'Outcome Subtype',
    'Animal Type',
    'Age upon Outcome',
    'Breed',
    'Color'
])

In [26]:
outcomes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132327 entries, 0 to 132326
Data columns (total 5 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   Animal ID         132327 non-null  object        
 1   Outcome Type      132302 non-null  object        
 2   Sex upon Outcome  132326 non-null  object        
 3   outcome_date      132327 non-null  datetime64[ns]
 4   age_in_days       132327 non-null  int64         
dtypes: datetime64[ns](1), int64(1), object(3)
memory usage: 5.0+ MB


In [27]:
outcomes.rename(columns = {'Outcome Type':'outcome_type'}, inplace = True)

In [28]:
#Now order the dataframe by Animal ID and intake_date.
#Then remove duplicates but leave the latest observation
outcomes = outcomes.sort_values(by = ['Animal ID', 'outcome_date'])

In [29]:
outcomes.head()

Animal ID     outcome_type Sex upon Outcome outcome_date  age_in_days
106962   A006100  Return to Owner    Neutered Male   2014-03-08         2434
52564    A006100  Return to Owner    Neutered Male   2014-12-20         2721
94209    A006100  Return to Owner    Neutered Male   2017-12-07         3804
36294    A047759         Transfer    Neutered Male   2014-04-07         3657
75671    A134067  Return to Owner    Neutered Male   2013-11-16         5875

In [30]:
#Drop duplicates
outcomes = outcomes.drop_duplicates(subset = 'Animal ID', keep = 'last', ignore_index = True)

In [31]:
outcomes.head()

Animal ID     outcome_type Sex upon Outcome outcome_date  age_in_days
0   A006100  Return to Owner    Neutered Male   2017-12-07         3804
1   A047759         Transfer    Neutered Male   2014-04-07         3657
2   A134067  Return to Owner    Neutered Male   2013-11-16         5875
3   A141142  Return to Owner    Spayed Female   2013-11-17         5648
4   A163459  Return to Owner    Intact Female   2014-11-14         5505

In [32]:
outcomes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 118342 entries, 0 to 118341
Data columns (total 5 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   Animal ID         118342 non-null  object        
 1   outcome_type      118321 non-null  object        
 2   Sex upon Outcome  118341 non-null  object        
 3   outcome_date      118342 non-null  datetime64[ns]
 4   age_in_days       118342 non-null  int64         
dtypes: datetime64[ns](1), int64(1), object(3)
memory usage: 4.5+ MB


### Merge the Intake and Outcome dataframes on Animal ID

In [33]:
animals = intakes.merge(right = outcomes, how = 'left', on = 'Animal ID')

In [34]:
animals.head()

Animal ID      intake_type intake_condition animal_type  \
0   A006100            Stray           Normal         Dog   
1   A047759  Owner Surrender           Normal         Dog   
2   A134067    Public Assist          Injured         Dog   
3   A141142            Stray             Aged         Dog   
4   A163459            Stray           Normal         Dog   

                         Breed         Color intake_date     outcome_type  \
0         Spinone Italiano Mix  Yellow/White  2017-12-07  Return to Owner   
1                    Dachshund      Tricolor  2014-04-02         Transfer   
2            Shetland Sheepdog   Brown/White  2013-11-16  Return to Owner   
3  Labrador Retriever/Pit Bull   Black/White  2013-11-16  Return to Owner   
4      Miniature Schnauzer Mix    Black/Gray  2014-11-14  Return to Owner   

  Sex upon Outcome outcome_date  age_in_days  
0    Neutered Male   2017-12-07       3804.0  
1    Neutered Male   2014-04-07       3657.0  
2    Neutered Male   2013-11-16       5875.0  
3    Spayed Female   2013-11-17       5648.0  
4    Intact Female   2014-11-14       5505.0

In [35]:
animals.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 118225 entries, 0 to 118224
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   Animal ID         118225 non-null  object        
 1   intake_type       118225 non-null  object        
 2   intake_condition  118225 non-null  object        
 3   animal_type       118225 non-null  object        
 4   Breed             118225 non-null  object        
 5   Color             118225 non-null  object        
 6   intake_date       118225 non-null  datetime64[ns]
 7   outcome_type      117507 non-null  object        
 8   Sex upon Outcome  117527 non-null  object        
 9   outcome_date      117528 non-null  datetime64[ns]
 10  age_in_days       117528 non-null  float64       
dtypes: datetime64[ns](2), float64(1), object(8)
memory usage: 10.8+ MB


In [36]:
animals.outcome_type.value_counts()

Adoption           51331
Transfer           37527
Return to Owner    17350
Euthanasia          8779
Died                1233
Disposal             607
Rto-Adopt            599
Missing               57
Relocate              24
Name: outcome_type, dtype: int64